s4587281 Improved Unet for 2018 ISICs Segmentation

ByronDowney opened the pull request:
Student Number: 45872815
Name: Byron Downey

Hi COMP3710 teaching team,

Please review my pull request, which includes an implementation of an Improved Unet model which aims to accurately segment the pre-processed ISICs 2018 dataset. The objective was to achieve a dice coefficient > 0.8 on the test dataset, however I could only achieve a dice coefficient of 0.7.

The Files Added are:
Driver.py
Model.py
README.md
Various .png images in the README_Images folder

More information is included in the README.md file, but if there are any questions or concerns, please feel free to share them.

Thanks,
Byron

ByronDowney committed:
initial commit to test

ByronDowney committed:
Added parameters to specify train/validation/test split. Changed approach to loading data, now loads directly into tensorflow datasets. Also outputs proper usage when called without enough arguments, or a train/validation/test split that does not add up to 1.

ByronDowney committed:
Now normalises the dataset. The previous commit sounds a bit strange because there was a commit before it that failed without me realising.

ByronDowney committed:
Finished creating the Improved Unet model. Not yet tested though.

ByronDowney committed:
Changed padding approach after realising the ISIC images are not all the same size. Currently google colab is crashing when trying to run even a very small model, so I plan to test it on a lab computer.

ByronDowney committed:
Fixed bug with the splitting of the dataset. Changed approach to resizing/padding again, now resizing everything to (324,324) because computer was running out of memory when images were too large e.g. (3354, 2230)

ByronDowney committed:
Fixed another bug with the splitting of the dataset after realising that it was loading as a BatchDataset. That also solved my problem of not being able to use a dice coefficient loss function. Model is now producing decent results, though not dice > 0.8 yet

ByronDowney committed:
Adjusted parameters, now calculates length of dataset within the program. Replaced the sigmoid activation function in second last layer with LeakyReLU. Improved commenting in model definition. Changed the learning rate of the adam optimizer to match the Improved Unet paper. Removed early stopping.

ByronDowney committed:
Moved most of the contents of Model.py to Driver.py. Model.py now only contains the keras structure of the model

ByronDowney committed:
Completed the README file. Contains information about the ISICS dataset, Unet, and some results.

ByronDowney committed:
Added most of the contents from Model.py, such as pre-processing, fitting, visualisation. Model.py now only contains the keras structure of the model

ByronDowney committed:
Committed Image files to be used by README

ByronDowney committed:
Fixed README formatting issues & removed comments

wooboyeong commented:
Code
* Reached 0.7 DSC
* Why do you have Dense layer at the end? - UNET is fully convolutional
* Might need to use color_mode='grayscale' when loading mask images
* Missing residual connection after context module

General Comments
* Commit messages OK
* Documentation OK
* Pull request OK

ByronDowney committed:
Updated the README file with new test Dice Coefficient result. Fixed some typos and leftover notes to myself.

ByronDowney committed:
Now loads mask images in grayscale.

ByronDowney committed:
Added residual connections after each context module. Removed unnecessary dense layer at the end of the network.

ByronDowney committed:
Updated image files to match changes to model structure and output after adding residual connections and removing dense layer.

ByronDowney committed:
Merged Formatting changes with updated README

ByronDowney committed:
Re-added files that were accidentally deleted while trying to fix issue with Git

ByronDowney commented:
Hi @wooboyeong,

Thank you for the feedback, I absolutely agree with all your suggested changes. I'm not sure why I included the Dense layer, I must have been half asleep when I put that in. I also completely overlooked the residual connections when looking through the improved Unet paper, so that is why they were missing. I also changed the color_mode to grayscale, but I don't know if that actually changed anything substantial. Regardless, I have implemented all your suggested changes. Additionally, I have made changes to the README.md and associated images to reflect the new structure of my model, and the output of that model.

I hope my work can now be merged into the repository, but if there are any further issues please feel free to let me know.

Thanks,
Byron

shakes76 commented:
## TF/Torch Usage
Adequate use and implementation, dense layer at the end is strange -1
## Good Practice (Design/Commenting)
Good spacing and comments
No Header blocks -1
## Algorithm 
Driver Script present
Shows Usage & Demo & Visualisation & Data usage 
Module present 
No Data leakage 
Difficulty: Easy -10
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No feedback required
Request Description OK
